### PR TITLE
adds IO#set_encoding_by_bom spec

### PR DESCRIPTION
**Pull request by moofkit**
Solving #745

> 
> Added IO#set_encoding_by_bom to check the BOM and set the external
> encoding. Bug #15210

### CODE REVIEW

File: core/io/set_encoding_by_bom_spec.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      File.write(@name, "abc")
+
+      @io.set_encoding_by_bom.should == nil
+    end
+
+    it 'returns exception if io not in binary mode' do
+      not_binary_io = new_io(@name, 'r')
+
+      -> { not_binary_io.set_encoding_by_bom }.should raise_error(ArgumentError, 'ASCII incompatible encoding needs binmode')
+      not_binary_io.close
```
Comment (eregon) [inline core/io/set_encoding_by_bom_spec.rb:33]: Could you move this in a begin/ensure/end? So even if the assertion fails, the IO is still closed.

File: core/io/set_encoding_by_bom_spec.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    rm_r @name
+  end
+
+  ruby_version_is "2.7" do
+    it "returns the result encoding if found BOM UTF-8 sequence" do
+      File.write(@name, "\u{FEFF}abc")
+
+      @io.set_encoding_by_bom.should == Encoding::UTF_8
+      @io.external_encoding.should == Encoding::UTF_8
+    end
```
Comment (eregon) [inline core/io/set_encoding_by_bom_spec.rb:21]: Could you also add examples for UTF-16LE and UTF16-BE BOM?

File: core/io/set_encoding_by_bom_spec.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  after :each do
+    @io.close unless @io.closed?
+    rm_r @name
+  end
+
+  ruby_version_is "2.7" do
+    it "returns the result encoding if found BOM UTF-8 sequence" do
+      File.write(@name, "\u{FEFF}abc")
```
Comment (eregon) [inline core/io/set_encoding_by_bom_spec.rb:17]: `File.binwrite` seems better here

**Review by eregon — COMMENTED** [review]
Thanks for the new specs!

**Review by eregon — APPROVED** [review]
Thank you, looks great.

**Comment by moofkit** [comment]
@eregon thanks for review!

**Pull request merged**